core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -322,6 +325,32 @@ public static long snapshotIdAsOfTime(Table table, long timestampMillis) {
     return snapshotId;
   }
 
+  /**
+   * Returns the ID of the most recent snapshot in the branch
+   *
+   * @param table a {@link Table}
+   * @param branch a {@link String}
+   * @param timestampMillis the timestamp in millis since the Unix epoch
+   * @return the snapshot ID
+   * @throws IllegalArgumentException when no snapshot is found in the table, on the given branch
+   * older than the timestamp
+   */
+  public static long snapshotIdAsOfTime(Table table, String branch, long timestampMillis) {
+    SnapshotRef ref = table.refs().get(branch);
+    Preconditions.checkArgument(ref != null, "Branch %s does not exist", branch);
+    Preconditions.checkArgument(ref.isBranch(), "Ref %s is a tag, not a branch", branch);
+    Long snapshotId = null;
+    for (Snapshot snapshot : ancestorsOf(ref.snapshotId(), table::snapshot)) {
+      if (snapshot.timestampMillis() <= timestampMillis) {
+        snapshotId = snapshot.snapshotId();
+      }
+    }
```

Comment:
I don't think this is right. Here we're iterating over the ancestors rather than the log entries like we do for time travel on the main branch. 

To be consistent with previous time travel behavior I think we need to leverage history logs but I think that means in the metadata we should actually maintain a history log for each branch (requiring a spec change). @rdblue @jackye1995 

Also we're not guaranteed timestamp ordering due to WAP commits, so I think the linear search is required as opposed to using a binary search. https://github.com/apache/iceberg/issues/3891